### PR TITLE
*: add API concurrency metrics

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref #7167 
should be merged after #7239

### What is changed and how does it work?
Enable concurrency monitoring for all APIs

```commit-message
```

### Check List

Tests

- Unit test
- Manual test (add detailed scripts or steps below)
image

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: metrics/grafana/pd.json [hunk 1]
```diff
@@ -8829,6 +8829,104 @@
...
             "alignLevel": null
           }
+        },
+        {
+          "aliasColors": {},
+          "bars": false,
+          "dashLength": 10,
+          "dashes": false,
+          "datasource": "${DS_TEST-CLUSTER}",
+          "description": "The concurrency number ofeach kind of gRPC commands",
```
Comment (rleungx) [inline metrics/grafana/pd.json:8839]:
```suggestion
          "description": "The concurrency number of each kind of gRPC commands",
```

File: pkg/ratelimit/concurrency_limiter.go [hunk 2]
```diff
@@ -26,11 +26,13 @@ func newConcurrencyLimiter(limit uint64) *concurrencyLimiter {
...
 }
 
+const unlimit = uint64(0)
+
 func (l *concurrencyLimiter) allow() bool {
 	l.mu.Lock()
 	defer l.mu.Unlock()
 
-	if l.current+1 <= l.limit {
+	if l.limit == unlimit || l.current+1 <= l.limit {
```
Comment (rleungx) [inline pkg/ratelimit/concurrency_limiter.go:35]: Why do we need it?

Comment (CabinfeverB) [reply pkg/ratelimit/concurrency_limiter.go:35]: Reusing `concurrencyLimiter` statistics, and add concurrencyLimiter to all APIs. So now when limit is equal to 0, there is no limit

Comment (rleungx) [reply pkg/ratelimit/concurrency_limiter.go:35]: if l.limit is equal to 0, this condition will always be false?

Comment (CabinfeverB) [reply pkg/ratelimit/concurrency_limiter.go:35]: if l.limit is equal to 0, we should always allow

File: server/grpc_service.go [hunk 3]
```diff
@@ -678,7 +744,7 @@ func (s *GrpcServer) GetStore(ctx context.Context, request *pdpb.GetStoreRequest
 	} else if rsp != nil {
 		return rsp.(*pdpb.GetStoreResponse), err
 	}
-
+	time.Sleep(120 * time.Millisecond)
```
Comment (rleungx) [inline server/grpc_service.go:747]: Why sleep?

Comment (CabinfeverB) [reply server/grpc_service.go:747]: Sorry, I forgot to delete this when addressing the comment. I can't count concurrency without this during manual testing.

File: pkg/ratelimit/controller.go [hunk 4]
```diff
@@ -27,12 +30,51 @@ type Controller struct {
...
+		select {
+		case <-l.ctx.Done():
+			return
+		case <-tricker.C:
+			l.limiters.Range(func(key, value any) bool {
+				limiter := value.(*limiter)
+				label := key.(string)
+				// Due to not in hot path, no need to save sub Gauge.
+				if con := limiter.getConcurrencyLimiter(); con != nil {
+					l.concurrencyGauge.WithLabelValues(l.apiType, label).Set(float64(con.getCurrent()))
```
Comment (rleungx) [inline pkg/ratelimit/controller.go:73]: Maybe we can find another way to show the statistics, the current implementation only captures the instant value every 15s.

File: pkg/ratelimit/concurrency_limiter.go [hunk 5]
```diff
@@ -68,3 +74,13 @@ func (l *concurrencyLimiter) getCurrent() uint64 {
 
 	return l.current
 }
+
+func (l *concurrencyLimiter) getMaxConcurrency() uint64 {
+	l.mu.RLock()
```
Comment (rleungx) [inline pkg/ratelimit/concurrency_limiter.go:79]: Should be `Lock`?

**Review by disksing — APPROVED** [review]

File: server/grpc_service.go (hunk 3 above)
Comment (nolouch) [reply server/grpc_service.go:747]: need remove it.

**Review by nolouch — COMMENTED** [review]
rest lgtm

**Review by rleungx — APPROVED** [review]

**Pull request merged**